src/validation-dispatcher/single-field-dispatcher.ts

```diff
@@ -3,128 +3,58 @@ import {
-  const fieldValidation: FieldValidation =
-    validationParams.fieldValidationCollection[currentIndex];
-
-  const validationFunction = isFunction(fieldValidation)
-    ? fieldValidation
-    : fieldValidation.validator;
-
-  validationFunction(
-    validationParams.value,
-    validationParams.values,
-    fieldValidation['customArgs'] // TODO: this can be a bit dirty, if function return undefined, if not the value
-  )
-    .then(fieldValidationResult => {
-      const validationResult = checkFieldValidationResult(
-        fieldValidationResult
-      );
-
-      const isLastValidatorInCollection = isLastIndexInArray(
-        currentIndex,
-        validationParams.fieldValidationCollection
-      );
-
-      switch (validationResult) {
-        case ValidationCheck.SUCCEEDED:
-          if (isLastValidatorInCollection) {
-            resolve(fieldValidationResult);
-          } else {
-            currentIndex++;
-            iterateFiringFieldValidations(
-              resolve,
-              reject,
-              validationParams,
-              currentIndex
-            );
-          }
-          break;
-        case ValidationCheck.ERROR_NOT_EXPECTED:
-          // show a console error, warn the user one of his validators is not well formed
-          // if last element just return a default succeed validation
-          // if not keep on iteraring
-          console.error(
-            'form-validators: One of the field validator is returning a non expected value.'
-          );
-          if (isLastValidatorInCollection) {
-            resolve(createDefaultValidationResult());
-          } else {
-            currentIndex++;
-            iterateFiringFieldValidations(
-              resolve,
-              reject,
-              validationParams,
-              currentIndex
-            );
-          }
-          break;
-        case ValidationCheck.FAILED:
-          resolve(fieldValidationResult);
-          break;
-      }
-    })
-    .catch(() => {
-      reject(currentIndex);
-    });
+const fireValidation = (
+  value,
+  values,
+  fieldValidation: FieldValidation
+): Promise<ValidationResult> => {
+  const validationFn = getValidationFn(fieldValidation);
+  return validationFn(value, values, fieldValidation['customArgs']).then(
+    checkValidationResult
+  );
 };
 
-export const fireSingleFieldValidations = (
-  values: any,
+const findFirstFailedValidation = (
   value: any,
+  values: any,
   fieldValidations: FieldValidation[]
-): Promise<ValidationResult> => {
-  const validationParams = createValidationParams(
-    values,
-    value,
-    fieldValidations
-  );
-
-  const fieldValidationResultPromise = new Promise<ValidationResult>(
-    (resolve, reject) => {
-      if (arrayContainsEntries(fieldValidations)) {
-        iterateFiringFieldValidations(resolve, reject, validationParams, 0);
-      } else {
-        resolve(createDefaultValidationResult());
-      }
-    }
+) =>
+  fieldValidations.reduce(
```

Comment:
Add this link to explain the chaining behavior: https://css-tricks.com/why-using-reduce-to-sequentially-resolve-promises-works/

Reply:
Done